### DISCUSSION TITLE
d2 integration for common dev tools such as Github/HackMD?

### DISCUSSION THREAD

**Discussion by andorsk** [terrastruct/d2#320] [General]
I've seen https://github.com/terrastruct/d2/issues/118 but what are the plans around getting a real native GH integration with d2 live and working? How would that work? 

I can say that it would make an **big** difference in utility for me if a codeblock such as ```d2``` was rendered the same way as a code block for ```mermaid``` in the following tools:

* github
* hackmd

I would love for this discussion to address the following:

* what additional tools to the ones listed would also be useful to integrate with? 
* what would be the process for getting a github/hackmd native render working, and how big of a gap would that be from today?

I'm putting this here b/c I'm not sure if there's been additional conversations outside of the issue list that discuss this topic, and not sure it has enough grounding yet to actually be posted as an issue.  I'm happy to migrate this to an issue if it makes sense. 

Some more context on why I **personally:** care about this. I tend to write a lot of charts in my markdown. I wrote an integration currently in review at melpa https://github.com/andorsk/d2-mode for emacs, so adding d2 and hackmd support would allow me to complete my flow of JIT rendering of d2 graphs for all my tools, which is a big deal and why I use mermaid so frequently in my documentation. I would love to have some alternatives, such as d2, but AFAIK, for GH, I would have to compile the render first then reference it in the markdown ( which I don't like ), or ask someone to clone the repo and then use a tool ( from the README for example ) to view the graph.

**Comment by alixander** [comment]
Sorry for the delay, I wrote a response a few days ago then my computer died and it got erased.

### GitHub integration

The bad news: the first hurdle to cross is getting enough D2 files in the public sphere for GitHub to even integrate it into their native syntax highlighting. This [requires](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#adding-an-extension-to-a-language) at least 200 public repos with .d2 files. After that, it's uncharted territory as far as I know, but there's some yet unknown level of demand to be met for GitHub to support natively rendering code blocks.

The good news: meanwhile, we're working on an API that lets you render images on the fly. Something like `<img src="https://render.d2lang.com?file=./docs/input.d2" />` and the API can find the file and serve it. We explored a GitHub app/bot that can also just parse GitHub comments and add the img directly, but they don't support adding configurations, e.g. theme. Writing this now, maybe we can allow specifying configs right in the script: https://github.com/terrastruct/d2/issues/410.

### HackMD

I hadn't heard of them before, I'll reach out. Perhaps they'll be more receptive to earlier stage tools.

d2-mode looks 💯 

Related
https://github.com/terrastruct/d2/discussions/345

**Reply by andorsk** [reply]
Thanks @alixander!  Sorry about the computer issue! I hate when that happens. Super helpful. Do we know how many people are using d2 files currently?

totally understand this is an ecosystem issue, but it's such an important one. I use mermaid all the time tbh :) if d2 wants to compete with mermaid and get the same levels of adoption, I think this is a really high priority to be honest, so I think this needs to be done. 

fyi: mermaid language.yml for reference: https://github.com/github/linguist/blob/712a1ee719675c98be4edd5337c71862c50c4e30/lib/linguist/languages.yml#L3985
